mypy/checker.py

```diff
@@ -2506,6 +2506,32 @@ def conditional_type_map(expr: Expression,
         return {}, {}
 
 
+def conditional_callable_type_map(expr: Expression,
+                                  current_type: Optional[Type],
+                                  ) -> Tuple[TypeMap, TypeMap]:
+    """Takes in an expression and the current type of the expression
+
+    Returns a 2-tuple: The first element is a map from the expression to
+    the restricted type if it must be callable. The second element is a
+    map from the expression to the type it would hold if it weren't
+    callable."""
+    if current_type:
+        if isinstance(current_type, CallableType):
+            # Expression is always of the type callable
+            return {}, None
+        elif isinstance(current_type, UnionType):
+            callables = [item for item in current_type.items
+                         if isinstance(item, CallableType)]  # type: List[Type]
+            non_callables = [item for item in current_type.items
+                             if not isinstance(item, CallableType)]  # type: List[Type]
+            return {expr: UnionType.make_union(callables)}, \
+                   {expr: UnionType.make_union(non_callables)}
+        else:
+            return None, {}
+    else:
+        return {}, {}
```

Comment:
nit: This could be factored with less indentation by reversing the order of conditional checks. Also, avoid backslashes. I know you based off `conditional_type_map()` but given it's already copied and pasted, how about:
```py3
    if not current_type:
        return {}, {}
        
    if isinstance(current_type, CallableType):
        return {}, None

    if isinstance(current_type, UnionType):
        callables = [item for item in current_type.items
                     if isinstance(item, CallableType)]  # type: List[Type]
        non_callables = [item for item in current_type.items
                         if not isinstance(item, CallableType)]  # type: List[Type]
        return (
            {expr: UnionType.make_union(callables)},
            {expr: UnionType.make_union(non_callables)},
        )

    return None, {}
```

The two list comprehensions are a bit wasteful but I guess that's fine, we don't expect unions to be overly long.